### PR TITLE
fix: role extraction from access token in keycloak oidc 

### PR DESCRIPTION
**Pull request by Elektordi**
The keycloak_oidc provider uses the access_token instead of the id_token.

## Description

I had 500 errors when using keycloak_oidc provider out of the box, and logs where indicating a problem with the audience:

```[oauthproxy.go:830] Error creating session during OAuth2 callback: audience claims [aud] do not exist in claims: [...]```

## Motivation and Context

oauth2-proxy relies on the audience "aud" field, and the officiel keycloak documentation states that the access_token is not design to hold the audience of the client requesting the token:

https://www.keycloak.org/docs/latest/server_admin/#_audience_resolve

> The frontend client itself is not automatically added to the access token audience, therefore allowing easy differentiation between the access token and the ID token, since the access token will not contain the client for which the token is issued as an audience.

It is the id_token which is designed for this, and this token is for the OIDC protocol. (where the access_token is for OAuth2 protocol)

It was also confirmed by Keycloak devs on their mailing list some time ago: https://lists.jboss.org/pipermail/keycloak-user/2019-August/018924.html

## How Has This Been Tested?

It has been tested with a brand now keycloak and oauth2-proxy, both deployed with docker, and using traefik/whoami as a backend.
Command line: `docker run --rm --name oauth2proxy -p 4180:4180 oauth2-proxy --provider=keycloak-oidc --oidc-issuer-url=http://172.17.0.1:8080/realms/test --email-domain=* --upstream=http://172.17.0.1:49156/ --reverse-proxy --http-address=0.0.0.0:4180 --client-id=dev-localhost --client-secret=xxxxxxxxxxxx --cookie-secret=xxxxxxxxxxxxxxxx --redirect-url=http://localhost:4180/oauth2/callback --skip-jwt-bearer-tokens --insecure-oidc-allow-unverified-email`

It was tested with both web access, and bearer api access.

Before change: HTTP 500 error on auth2-proxy

After change : No problem.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

(I'm not sure if I need to update "Breaking Changes" in CHANGELOG, because it may break installations where people have customized their KC to add specific stuff to their access_token, even is the specs says you should not)

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Have you looked at the history of this provider? Is it possible the behaviour has changed in keycloak and your change makes it supported for newer versions but breaks older versions?
Would be good to get reviews and acks from some other keycloak folk before we merge.

If you can get a keycloak maintainer to comment that would help as well

**Comment by Elektordi** [comment]
One of the devs from redhat already confirmed this:
https://lists.jboss.org/pipermail/keycloak-user/2019-August/018924.html
I'll check for changes about this in history of the file..

**Comment by babs** [comment]
IMHO, this is still relevant.

**Comment by JoelSpeed** [comment]
This sounds like something we should fix, but I'm wondering if there's a way to avoid a breaking change here, any thoughts?

**Comment by babs** [comment]
I'm not quite sure what it could break as the later adds aud claims defacto among other things if I remember correctly.
If users want to use the access token, the proper provider to use is the non oidc flavor of keycloak ;)

**Comment by yann-soubeyrand** [comment]
Hello, I’v asked Keycloak developers for some clarification:

**Comment by babs** [comment]
As stated by @Elektordi , 

> access_token is from Oauth2 spec (and used in keycloak provider), and id_token is from oidc spec.

If you want to use access_token you can use the `keycloak` provider, and leave `keycloak-oidc` for .... oidc and therefore use the proper token :)

**Comment by yann-soubeyrand** [comment]
> As stated by @Elektordi ,
> 
> > access_token is from Oauth2 spec (and used in keycloak provider), and id_token is from oidc spec.

Yes, I know what Oauth2 and OIDC are (technically OIDC is based on Oauth2, so the access token is also and OIDC thing, but I won’t nitpick on this).

> 
> If you want to use access_token you can use the `keycloak` provider, and leave `keycloak-oidc` for .... oidc and therefore use the proper token :)

Well, if you want pure OIDC, I guess you can just use the OpenID Connect provider. I’m sure you’re aware that the Keycloak provider is already making use of the access token to get the roles and put them in the group list :wink:

https://github.com/oauth2-proxy/oauth2-proxy/blob/b78c391adc6da2598870f19469d18ca1ab035897/providers/keycloak_oidc.go#L109-L121

https://github.com/oauth2-proxy/oauth2-proxy/blob/b78c391adc6da2598870f19469d18ca1ab035897/providers/keycloak_oidc.go#L142-L159

**Comment by babs** [comment]
> > If you want to use access_token you can use the `keycloak` provider, and leave `keycloak-oidc` for .... oidc and therefore use the proper token :)
> 
> Well, if you want pure OIDC, I guess you can just use the OpenID Connect provider. I’m sure you’re aware that the Keycloak provider is already making use of the access token to get the roles and put them in the group list 😉
> 

@yann-soubeyrand yes, I know, but in this case the `aud` is missing or wrong, so an extra config has to be made on KC side to make it work comply with OIDC (that, I can't emphase enought, shouldn't have to be done if proper token was used), all that because oauth2-proxy doesn't use the proper token due to legacy KC issue and a global misunderstanding.

I'd also argue: why having 2 code bases to talk to keycloak using the exact same token while the very purpose of `keycloak-oidc` should be to be defacto compliant with the specs.
It's not like we are talking about removing functionality here, IMHO, it's more fixing a legacy misuse and/or misunderstanding and making things right.

**Comment by yann-soubeyrand** [comment]
The problem is that, by default in Keycloak, no ID token claim contains the roles. So it seems there’s no way to avoid user configuration and to exclusively use the ID token. That’s why I asked on Keycloak discussions an advice on the way to go, but I got no answer so far.

**Comment by Elektordi** [comment]
> no ID token claim contains the roles
I agree, but roles is a more specific feature, and not a needed one (you can use OIDC without roles).

> why having 2 code bases to talk to keycloak using the exact same token while the very purpose of keycloak-oidc should be to be defacto compliant with the specs.
That is what I was thinking, the `keycloak` provider should use access token, and `keycloak-oidc` the id token, and things would work out of the box (without roles, indeed).

And, btw, multiple people have opened issues because of this bug.
For example, last one I found: https://github.com/oauth2-proxy/oauth2-proxy/issues/2117

**Comment by yann-soubeyrand** [comment]
OK, but then what’s the added value of the Keycloak OIDC provider compared to the generic OIDC provider?

**Comment by babs** [comment]
good question, initially probably not that much, now there is another pull request for backend logout for example

**Comment by babs** [comment]
Any update @JoelSpeed ?

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,8 +6,11 @@
 
 ## Breaking Changes
 
+- [#1916](https://github.com/oauth2-proxy/oauth2-proxy/pull/1916) Keycloak OIDC provider was using access token instead of id token. After upgrade, it may raise errors on login for people using provider "keycloak-oidc" (not "keycloak") and having customized their tokens in keycloak configuration.
```
[the remark this answers is not in the document]
Comment (Elektordi) [reply CHANGELOG.md:9]: Yes, I'm still here! Yes, I'm still motivated.
I'll make some tests, but I think that people who have already customized configuration should know how to remove it. And if they only add new fields, the change should not affect them since they do not change the existing fields.
But I can make it more clear if you think it can be an issue!

[the remark this answers is not in the document]
Comment (babs) [reply CHANGELOG.md:9]: @Elektordi @kvanzuijlen any news on this topic ?

**Comment by babs** [comment]
Not stalled

**Comment by wearpants** [comment]
Can confirm, this problem still exists, I worked around it with a custom mapper in keycloak, as per https://github.com/oauth2-proxy/oauth2-proxy/issues/2117#issue-1712501553

**Comment by babs** [comment]
It works out if the bix without the custom mapper with this patch. I honestly think having somthing working out of the box should be the default behavior, even more when it specifically states keycloak in the name ;)

**Comment by whg517** [comment]
Sorry guys, I've seen this problem for a long time. I am using the integration of the oauth-proxy authentication scheme on k8s. I also encountered this problem when using keycloak-oidc, referring to the documentation operation. I'm no expert on keycloak, so does the core team have a specific strategy for this issue? Whether it's improving oauth2-proxy, or refining a more complex configuration of keycloak through documentation? If you have any news, or a roadmap, please let everyone know.

**Comment by babs** [comment]
@kvanzuijlen @Elektordi ping ?

**Comment by olivierlambert** [comment]
I had this temporary fix in Keycloak for a while, and now with Keycloak 25.1, I can't make it work anymore for any new client I add (or if I modify a previous client):

```
[oauthproxy.go:888] Error creating session during OAuth2 callback: audience from claim aud with value [xxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx account] does not match with any of allowed audiences map[<client-id>:{}]
```

I checked 10 times, follow the doc in the exact same details and despite that, I can't make it work anymore.

**Comment by PaddyKe** [comment]
> Does anyone with a recent Keycloak made it work?
> 
> That would be great to get it solved via this PR, so we do not need any extra config anymore

I just ran into the same issue. For some reason, Keycloak 25 seems to add some kind of GUID into the aud scope instead of the actual client id.
I was able to make it work by not specifying any value in the "Include Client Audience" and instead manually entering the client id in "Include Custom Audience" field.

image

**Comment by babs** [comment]
@Elektordi @JoelSpeed @kvanzuijlen little up ?

**Comment by Fmstrat** [comment]
Could someone provide a how-to for the workaround?

**Comment by JoniJnm** [comment]
I "fixed" it using [oidc_extra_audiences](https://oauth2-proxy.github.io/oauth2-proxy/configuration/overview#general-provider-options) with a fixed value. Maybe `aud` or `account`, check the jwt token

```
oidc_extra_audiences = [ "account" ]
```

**Comment by babs** [comment]
@JoelSpeed  ?

**Comment by franco-viotti** [comment]
@PaddyKe solution worked for me, thanks! I'm using Keycloak 25.0.1 and oauth2-proxy 7.6.0

**Comment by tuunit** [comment]
Hey guys,

as this has been open for years now and I fully agree with @babs on the following:

> I'd also argue: why having 2 code bases to talk to keycloak using the exact same token while the very purpose of keycloak-oidc should be to be defacto compliant with the specs.
It's not like we are talking about removing functionality here, IMHO, it's more fixing a legacy misuse and/or misunderstanding and making things right.

I investigated this whole mess in-depth!

First of all, I want to get some confusion out of the way.

> OK, but then what’s the added value of the Keycloak OIDC provider compared to the generic OIDC provider?

TLDR; The Keycloak provider is deprecated and the KeycloakOIDC provider extends the generic OIDC provider and therefore offers all generic OIDC options and on top the extraction of roles from Keycloaks API.

A couple years ago, Nick and Joel started working on generalising the code base. One of the goals was and still is, to try to have at little as possible specific provider code as possible, for this reason the generic OIDC provider was introduced and in the long run all OIDC compliant providers should only extend its functionality. The Keycloak provider was kept to stay backwards compatible and not break existing users setup. We now highlight in the docs and in the provider code that the normal Keycloak provider is therefore deprecated and people should use the KeycloakOIDC provider.



## Main point which this PR was supposed to address:

**Standard Compliance**
Align with the OIDC specification, where the ID token is meant for authentication (containing user identity claims) and Access tokens are for authorization (OAuth2) and **aren't** guaranteed to contain the expected `aud` claim.

**Breaking change**
As already identified, the code was build under the assumption that the access token is used for extracting the roles:

https://github.com/oauth2-proxy/oauth2-proxy/blob/367183d7b8d82be2fba2722c79e469cafdec1418/providers/keycloak_oidc.go#L100-L107

https://github.com/oauth2-proxy/oauth2-proxy/blob/367183d7b8d82be2fba2722c79e469cafdec1418/providers/keycloak_oidc.go#L83-L98

Those aren't included in the ID Token returned by Keycloak:

ID Token:
```
{
  "exp": 1745780789,
  "iat": 1745780729,
  "auth_time": 1745780729,
  "jti": "76d39596-4af7-4bf5-8688-72232e064a42",
  "iss": "http://keycloak.localtest.me:9080/realms/oauth2-proxy",
  "aud": "oauth2-proxy",
  "sub": "3356c0a0-d4d5-4436-9c5a-2299c71c08ec",
  "typ": "ID",
  "azp": "oauth2-proxy",
  "sid": "8eb0bed4-de02-46e7-824f-6bb8d476a590",
  "at_hash": "pr_FFqbZrJ4q7SYld2B5Cg",
  "email_verified": true,
  "preferred_username": "admin@example.com",
  "email": "admin@example.com"
}
```

Access Token:
```
{                                                                                                                                                                                                                                               
  "exp": 1745780789,                                                                                                                                                                                                                            
  "iat": 1745780729,                                                                                                                                                                                                                            
  "auth_time": 1745780729,                                                                                                                                                                                                                      
  "jti": "af7fb1dd-143e-4d83-80f5-bb4201b2c322",                                                                                                                                                                                                
  "iss": "http://keycloak.localtest.me:9080/realms/oauth2-proxy",                                                                                                                                                                               
  "aud": [                                                                                                                                                                                                                                      
    "oauth2-proxy-realm",                                                                                                                                                                                                                       
    "account"                                                                                                                                                                                                                                   
  ],                                                                                                                                                                                                                                            
  "sub": "3356c0a0-d4d5-4436-9c5a-2299c71c08ec",                                                                                                                                                                                                
  "typ": "Bearer",                                                                                                                                                                                                                              
  "azp": "oauth2-proxy",                                                                                                                                                                                                                        
  "sid": "8eb0bed4-de02-46e7-824f-6bb8d476a590",                                                                                                                                                                                                
  "realm_access": {                                                                                                                                                                                                                             
    "roles": [                                                                                                                                                                                                                                  
      "create-realm",                                                                                                                                                                                                                           
      "offline_access",                                                                                                                                                                                                                         
      "admin",                                                                                                                                                                                                                                  
      "uma_authorization"                                                                                                                                                                       
    ]                                                                                                                                                                                           
  },                                                                                                                                                                                            
  "resource_access": {                                                                                                                                                                          
    "oauth2-proxy-realm": {                                                                                                                                                                     
      "roles": [                                                                                                                                                                                
        "view-realm",                                                                                                                                                                           
        "view-identity-providers",                                                                                                                                                              
        "manage-identity-providers",                                                                                                                                                            
        "impersonation",                                                                                                                                                                        
        "create-client",                                                                                                                                                                        
        "manage-users",                                                                                                                                                                         
        "query-realms",                                                                                                                                                                         
        "view-authorization",                                                                                                                                                                   
        "query-clients",                                                                                                                                                                        
        "query-users",                                                                                                                                                                          
        "manage-events",                                                                                                                                                                        
        "manage-realm",                                                                                                                                                                         
        "view-events",                                                                                                                                                                          
        "view-users",                                                                                                                                                                           
        "view-clients",                                                                                                                                                                         
        "manage-authorization",                                                                                                                                                                 
        "manage-clients",                                                                                                                                                                       
        "query-groups"                                                                                                                                                                          
      ]                                                                                                                                                                                         
    },                                          
    "account": {                                
      "roles": [                                
        "manage-account",                       
        "manage-account-links",                 
        "view-profile"                          
      ]                                         
    }                                           
  },                                            
  "scope": "openid profile email",              
  "email_verified": true,                       
  "preferred_username": "admin@example.com",    
  "email": "admin@example.com"                  
}
```

## Drill down

As `keycloak_oidc.go` is inheriting the generic functionality from `oidc.go` the actual session validation and creation is happening here:

https://github.com/oauth2-proxy/oauth2-proxy/blob/367183d7b8d82be2fba2722c79e469cafdec1418/providers/oidc.go#L222-L250

https://github.com/oauth2-proxy/oauth2-proxy/blob/367183d7b8d82be2fba2722c79e469cafdec1418/providers/oidc.go#L109-L115

And those locations are already properly using the `access_token` and `id_token` respectively. 
Therefore this PR doesn't actually address the OIDC compliance as it is already in place and properly working.

**So what is the problem then?**
What this PR actually changed isn't the session creation or the session validation logic but instead it is just fighting a symptom of a strange implementation. By which we abuse the `p.Verifier.Verify` method to extract an interface / object from the `access_token` of type string to be able to call the claim extraction method on it.

What really needs to be fixed is how we extract the claims from the access token of Keycloak which at the point of usage in `getAccessClaims(...)` has already been through the exchange and validation. 

Therefore I propose we remove the unnecessary `verification` in:

https://github.com/oauth2-proxy/oauth2-proxy/blob/367183d7b8d82be2fba2722c79e469cafdec1418/providers/keycloak_oidc.go#L109-L114

and instead fix the actual problem like so:

```
func (p *KeycloakOIDCProvider) getAccessClaims(ctx context.Context, s *sessions.SessionState) (*accessClaims, error) {
	var claims *accessClaims
	if err := json.Unmarshal([]byte(s.AccessToken), claims); err != nil {
		return nil, err
	}
	return claims, nil
}
```

What you correctly identified is the fact that the tests did indeed use the Access Token instead of the ID Token which was an an oversight and should stay as proposed in this PR.

CC: @JoelSpeed @Elektordi @babs

**Comment by tuunit** [comment]
I updated the branch. Applied my suggested fix and fixed the tests accordingly.

**Review by tuunit — APPROVED** [review]

**Review by tuunit — APPROVED** [review]
